### DISCUSSION TITLE
How to retry failed request with the same query keys

### DISCUSSION THREAD

**Discussion by donysukardi** [TanStack/query#2702] [Q&A]
Suppose we have a setup like the following, with SearchForm and SearchResults components that operates on searchParams state.

![Search Page](https://user-images.githubusercontent.com/410792/134316650-429ee33f-3a93-4b1c-95af-69aa94b134ec.png)

With such setup, when useQuery fails and user presses Submit button again, the useQuery won't rerun since the `searchParams` is deep equal with the previous value.

Is there any recommended pattern to solve this?

I can think of one solution, which is to compare new `searchParams` with existing one and if it's deep equal, I invalidate existing query instead.

**Comment by TkDodo** [comment]
I usually call `refetch` if the params are equal. You can also do refetch if the query is erroneous

**Reply by donysukardi** [reply]
To do that, the useQuery hook has to be moved up to the parent component, doesn't it?

**Reply by TkDodo** [reply]
You can also refetch via the queryClient
